network/topics.go

```diff
@@ -0,0 +1,132 @@
+// Copyright (C) 2019-2020 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package network
+
+import (
+	"encoding/binary"
+	"fmt"
+
+	"github.com/algorand/go-algorand/crypto"
+)
+
+// Topic is a key-value pair
+type Topic struct {
+	key  string
+	data []byte
+}
+
+// Topics is an array of type Topic
+// The maximum number of topics allowed is 32
+// Each topic key can be 64 characters long and cannot be size 0
+type Topics []Topic
+
+// MarshallTopics serializes the topics into a byte array
+func (ts Topics) MarshallTopics() (b []byte) {
+
+	// Calculate the total buffer size required to store the topics
+	bufferSize := binary.MaxVarintLen32 // store topic array size
```

Comment:
should not it be binary. MaxVarintLen64 since uint64(len(ts)) is written below?

Reply:
bufferSize is estimating the buffer size, and on this line, it is adding the needed space for storing the topic array size. 
The maximum topic array size can be 32, and the estimate here is for storing a 32-bit integer. So, we are not under-estimating here. On the contrary, over-estimating.
Below, it is converted to uint64, because the function expects uint64, and not because it may need 64bits to store the value. 

The only situation were the code is not handling is when the number of topics passed is greater than or equal to 2^32. This will be an error when unmarshalling the topics.

Reply:
If we ever get to a situation where we trying to send 2GB of data in a single "message", than we have a serious design issue here.. sure we can test it here, but the source for the issue wouldn't be here but rather in the application level.

Reply:
@algonautshant, while I don't think it would never be an issue, could you change this into  MaxVarintLen64, or change the `uint64(len(ts))` into `uint32(len(ts))` ?

Reply:
My bad, I misread the code